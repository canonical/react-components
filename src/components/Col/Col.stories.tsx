import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import Col from "./Col";
import Row from "../Row";

const Template = (args) => {
  return (
    <div className="grid-demo">
      <Row>
        <Col {...args} />
      </Row>
    </div>
  );
};

const meta: Meta<typeof Template> = {
  component: Col,
  render: Template,
  tags: ["autodocs"],

  argTypes: {
    children: {
      control: {
        type: "text",
      },
    },

    size: {
      control: {
        type: "range",
        min: 1,
        max: 12,
        step: 1,
      },
    },
  },
};

export default meta;

type Story = StoryObj<typeof Col>;

export const Default: Story = {
  name: "Default",

  args: {
    children: "Hi, I'm Col.",
    size: 12,
  },
};

export const Grid: Story = {
  render: () => {
    return (
      <div className="grid-demo">
        <Row>
          <Col size={12}>
            <span>.col-12</span>
          </Col>
        </Row>
        <Row>
          <Col size={11}>
            <span>.col-11</span>
          </Col>
          <Col size={1}>
            <span>.col-1</span>
          </Col>
        </Row>
        <Row>
          <Col size={10}>
            <span>.col-10</span>
          </Col>
          <Col size={2}>
            <span>.col-2</span>
          </Col>
        </Row>
        <Row>
          <Col size={9}>
            <span>.col-9</span>
          </Col>
          <Col size={3}>
            <span>.col-3</span>
          </Col>
        </Row>
        <Row>
          <Col size={8}>
            <span>.col-8</span>
          </Col>
          <Col size={4}>
            <span>.col-4</span>
          </Col>
        </Row>
        <Row>
          <Col size={7}>
            <span>.col-7</span>
          </Col>
          <Col size={5}>
            <span>.col-5</span>
          </Col>
        </Row>
        <Row>
          <Col size={6}>
            <span>.col-6</span>
          </Col>
          <Col size={6}>
            <span>.col-6</span>
          </Col>
        </Row>
        <Row>
          <Col size={5}>
            <span>.col-5</span>
          </Col>
          <Col size={7}>
            <span>.col-7</span>
          </Col>
        </Row>
        <Row>
          <Col size={4}>
            <span>.col-4</span>
          </Col>
          <Col size={8}>
            <span>.col-8</span>
          </Col>
        </Row>
        <Row>
          <Col size={3}>
            <span>.col-3</span>
          </Col>
          <Col size={9}>
            <span>.col-9</span>
          </Col>
        </Row>
        <Row>
          <Col size={2}>
            <span>.col-2</span>
          </Col>
          <Col size={10}>
            <span>.col-10</span>
          </Col>
        </Row>
        <Row>
          <Col size={1}>
            <span>.col-1</span>
          </Col>
          <Col size={11}>
            <span>.col-11</span>
          </Col>
        </Row>
      </div>
    );
  },

  name: "Grid",
};

export const NestedColumns: Story = {
  render: () => {
    return (
      <div className="grid-demo">
        <Row>
          <Col small={4} medium={6} size={12}>
            col-small-4 .col-medium-6 .col-12
            <Row>
              <Col small={3} medium={3} size={9}>
                col-small-3 .col-medium-3 .col-9
                <Row>
                  <Col small={1} medium={1} size={2}>
                    col-small-1 col-medium-1 col-2
                  </Col>
                  <Col small={1} medium={1} size={3}>
                    col-small-1 col-medium-1 col-2
                  </Col>
                  <Col small={1} medium={1} size={3}>
                    col-small-1 col-medium-1 col-2
                  </Col>
                </Row>
              </Col>
              <Col small={1} medium={3} size={3}>
                col-small-2 col-medium-3 col-3
              </Col>
            </Row>
          </Col>
        </Row>
      </div>
    );
  },

  name: "Nested columns",
};

export const EmptyColumns: Story = {
  render: () => {
    return (
      <div className="grid-demo">
        <Row>
          <Col size={8}>.col-8</Col>
          <Col size={4}>.col-4</Col>
        </Row>
        <Row>
          <Col size={7}>.col-7</Col>
          <Col size={4}>
            <Row>
              <Col size={3} emptyLarge={2}>
                col-3 col-start-large-2 inside col-4
              </Col>
            </Row>
          </Col>
        </Row>
        <Row>
          <Col size={7}>.col-7</Col>
          <Col size={4}>
            <Row>
              <Col size={3}>col-3 inside col-4</Col>
            </Row>
          </Col>
        </Row>
      </div>
    );
  },

  name: "Empty columns",
};
