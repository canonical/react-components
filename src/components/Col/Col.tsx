import classNames from "classnames";
import React from "react";
import type { ElementType, HTMLProps, ReactNode } from "react";

import type { ClassName, PropsWithSpread } from "types";

export type ColSize = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;

export const colSizes: ColSize[] = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12];

export type Props = PropsWithSpread<
  {
    /**
     * The content of the column.
     */
    children?: ReactNode;
    /**
     * Optional class(es) to pass to the wrapping element.
     */
    className?: ClassName;
    /**
     * Optional element type to give the wrapper if not "div".
     */
    element?: ElementType;
    /**
     * The number of columns to skip before starting on large screens.
     */
    emptyLarge?: ColSize;
    /**
     * The number of columns to skip before starting on medium screens.
     */
    emptyMedium?: ColSize;
    /**
     * The number of columns to skip before starting on small screens.
     */
    emptySmall?: ColSize;
    /**
     * Override for the number of columns the content occupies on large screens.
     */
    large?: ColSize;
    /**
     * Override for the number of columns the content occupies on medium screens.
     */
    medium?: ColSize;
    /**
     * The number of columns the content occupies.
     */
    size: ColSize;
    /**
     * Override for the number of columns the content occupies on small screens.
     */
    small?: ColSize;
  },
  HTMLProps<HTMLElement>
>;

/**
 * This is a [React](https://reactjs.org/) component for use within the Vanilla [Grid](https://docs.vanillaframework.io/patterns/grid/).
 *
 * Vanilla has a responsive grid using a combination of rows and columns.
 */
const Col = ({
  children,
  className,
  element: Component = "div",
  emptyLarge,
  emptyMedium,
  emptySmall,
  large,
  medium,
  size,
  small,
  ...props
}: Props): JSX.Element => (
  <Component
    className={classNames(className, `col-${size}`, {
      [`col-small-${small}`]: !!small,
      [`col-medium-${medium}`]: !!medium,
      [`col-large-${large}`]: !!large,
      [`col-start-small-${emptySmall}`]: !!emptySmall,
      [`col-start-medium-${emptyMedium}`]: !!emptyMedium,
      [`col-start-large-${emptyLarge}`]: !!emptyLarge,
    })}
    {...props}
  >
    {children}
  </Component>
);

export default Col;
