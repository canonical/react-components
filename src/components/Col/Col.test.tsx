import { render, screen } from "@testing-library/react";
import React from "react";

import Col from "./Col";

describe("Col ", () => {
  it("renders", () => {
    render(
      <Col data-testid="col" size={3}>
        Test content
      </Col>
    );
    expect(screen.getByTestId("col")).toMatchSnapshot();
  });

  it("can set a size for small screens", () => {
    render(
      <Col data-testid="col" size={3} small={1}>
        Test content
      </Col>
    );
    expect(screen.getByTestId("col")).toHaveClass("col-small-1");
  });

  it("can set a size for medium screens", () => {
    render(
      <Col data-testid="col" size={3} medium={1}>
        Test content
      </Col>
    );
    expect(screen.getByTestId("col")).toHaveClass("col-medium-1");
  });

  it("can set a size for large screens", () => {
    render(
      <Col data-testid="col" size={3} large={1}>
        Test content
      </Col>
    );
    expect(screen.getByTestId("col")).toHaveClass("col-large-1");
  });

  it("can set empty cols for small screens", () => {
    render(
      <Col data-testid="col" size={3} emptySmall={1}>
        Test content
      </Col>
    );
    expect(screen.getByTestId("col")).toHaveClass("col-start-small-1");
  });

  it("can set empty cols for medium screens", () => {
    render(
      <Col data-testid="col" size={3} emptyMedium={1}>
        Test content
      </Col>
    );
    expect(screen.getByTestId("col")).toHaveClass("col-start-medium-1");
  });

  it("can set empty cols for large screens", () => {
    render(
      <Col data-testid="col" size={3} emptyLarge={1}>
        Test content
      </Col>
    );
    expect(screen.getByTestId("col")).toHaveClass("col-start-large-1");
  });

  it("can add additional classes", () => {
    render(<Col data-testid="col" size={3} className="extra-class"></Col>);
    const col = screen.getByTestId("col");
    expect(col).toHaveClass("col-3");
    expect(col).toHaveClass("extra-class");
  });
});
