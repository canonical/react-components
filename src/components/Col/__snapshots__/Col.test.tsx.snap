// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Col  renders 1`] = `
<div
  class="col-3"
  data-testid="col"
>
  Test content
</div>
`;
